pkg/ioutils/readers.go

```diff
@@ -74,14 +93,61 @@ func NewBufReaderWithDrainbufAndBuffer(r io.Reader, drainBuffer []byte, buffer *
 }
 
 func (r *bufReader) drain() {
+	var (
+		duration       time.Duration
+		lastReset      time.Time
+		now            time.Time
+		reset          bool
+		bufLen         int64
+		dataSinceReset int64
+		maxBufLen      int64
+		reuseBufLen    int64
+		reuseCount     int64
+	)
+	reuseBufLen = int64(len(r.reuseBuf))
+	lastReset = time.Now()
 	for {
 		n, err := r.reader.Read(r.drainBuf)
+		dataSinceReset += int64(n)
 		r.Lock()
+		bufLen = int64(r.buf.Len())
+		if bufLen > maxBufLen {
+			maxBufLen = int64(bufLen)
+		}
+		if bufLen > 0 && reuseBufLen >= bufLen {
+			n, _ := r.buf.Read(r.reuseBuf)
+			r.buf.Reset()
+			r.buf.Write(r.reuseBuf[0:n])
+		} else if reuseCount >= r.maxReuse && bufLen > 0 {
+			now = time.Now()
+			duration = now.Sub(lastReset)
+			timeoutReached := duration >= r.resetTimeout
+			if timeoutReached && bufLen > reuseBufLen {
+				reset = true
+			}
+			if timeoutReached && maxBufLen >= r.bufLenResetThreshold {
+				reset = true
+			}
+			if timeoutReached && dataSinceReset >= r.maxReadDataReset {
```

Comment:
For more readability, we could define these functions at the top of `drain()`, after declaring all of the `var`s:
```go
timeoutReached = func() bool { return time.Now().Sub(lastReset) >= r.resetTimeout }
resetCond1 = func() bool { return bufLen > reuseBufLen }
resetCond2 = func() bool { return maxBufLen >= r.bufLenResetThreshold }
resetCond3 = func() bool { return dataSinceReset >= r.maxReadDataReset }
shouldReset = func() bool { return timeoutReached() && (resetCond1() || resetCond2() || resetCond3()) }
```
Then you can say:
```go
if shouldReset() {
	newbuf := &bytes.Buffer{}
	newbuf.ReadFrom(r.buf)
	r.buf = newbuf
	lastReset = time.Now()
	dataSinceReset = 0
	maxBufLen = 0
	reuseCount = 0
}
```